### ISSUE TITLE
clipdel: Delete current clip

### ISSUE THREAD

**Issue by brunelli** [cdown/clipmenu#98]
I'd like to have a feature in `clipdel` to delete the current clip (first item of `clipmenu`), so I wrote [a WIP patch](https://brunelli.me/dl/clipdel-delete-last.patch) which deletes the current clip and sets the clipboard content to the next item. It currently works with `clipdel -l`.

I'm opening this issue so we can discuss what's the best way to implement this feature.

**Comment by GladOSkar** [comment]
While i think the use case is valid, i think it would make more sense to create a "Delete picker" (with `dmenu`) as described in #57 since that would offer almost the same functionality.

**Comment by brunelli** [comment]
Not really, since I wouldn't be able to delete the last clip by just invoking a shortcut combo.

The issue gave me some ideas to implement this together with a clipdelmenu, though. I'll try to hack something when I get some time to spare.

**Comment by GladOSkar** [comment]
Yeah i guess ideally you could pass a number as a parameter and if no parameter is passed it asks interactively.

For the interactive mode one could even (for rofi users) use rofi's multi-select.

**Comment by brunelli** [comment]
> Yeah i guess ideally you could pass a number as a parameter for which item to delete and if no parameter is passed it asks interactively.

I'm not sure about that, as there's no indication of which number corresponds to an entry.

> For the interactive mode one could even (for rofi users) use rofi's multi-select.

Yeap, that'll totally be possible.

**Comment by GladOSkar** [comment]
> I'm not sure about that, as there's no indication of which number corresponds to an entry.

Well i would say in the order of the history `1` being the last item copied, `2` the one before that etc.

**Comment by brunelli** [comment]
> Well i would say in the order of the history: 1 being the last item copied, 2 the one before that etc.

Sure, but what's the use case?

The only way to know a entry's position is by running `clip{,del}menu` interactively, and even there there's no way to tell what's the associated number without counting.

**Comment by GladOSkar** [comment]
I only thought of your use case: deleting the last element

**Comment by brunelli** [comment]
I was actually thinking of implementing it like `clipdel --current`.

But don't get me wrong, I'm all in for something like `clipdel --entries 1 3-6`, but I'm not sure that would be useful with the way `clipmenu` works.

@cdown have you thought of adding numbers in front of the entries in `clipmenu`?

Something like:
```
------------------------------------------------------
| clipmenu:                                          |
------------------------------------------------------
| [1] First entry                                    |
| [2] Second entry                                   |
| [3] A very very very very long entry... (12 lines) |
| [4] Lorem ipsum                                    |
| [5] Dolor sit amet                                 |
------------------------------------------------------
```

**Comment by GladOSkar** [comment]
I guess you're right, i don't really see a reason to automate deleting any other but the last copied entry

**Comment by ertwro** [comment]
I use something like this in the clipmenu script when I start it with a flag but I agree is not necessary.
```bash
rm_line=$(grep -Fn "$chosen_line" "$cache_file" | cut -d':' -f1 | sed 's/$/d/' | paste -sd';')
sed -i "${rm_line}" "$cache_file"
rm -f "$file"
exit 0
```

**Comment by OJFord** [comment]
> I guess you're right, i don't really see a reason to automate deleting any other but the last copied entry

Passwords, copied to clipboard on a timeout (for example, `pass` defaults to 45s) - you might copy something else in that time; the password manager wants to delete a known value, wherever it appears in order.

**Comment by snprajwal** [comment]
@cdown bumping this up since I too use `pass` with `clipmenu`, and don't want my password lingering in the clipboard. Would you accept a PR for this?

**Comment by cdown** [comment]
@snprajwal You can disable clipmenud's collection like so when using pass:

```
clipctl disable
[... do something ...]
clipctl enable
```

**Comment by cdown** [comment]
As for this general issue, it seems somewhat related to #57, which should be trivial to implement in the current version. PRs welcome, of course :)

**Comment by N-R-K** [comment]
Is this (or #57 for that matter) even needed anymore? Skimming through the thread it seems like the use case was to avoid storing password but we already have `CM_IGNORE_WINDOW` and also `clipctl disable`.

**Comment by snprajwal** [comment]
`CM_IGNORE_WINDOW` would not work for `pass` because it does not create a window at all - it simply runs in the terminal you invoke it in. And creating a shell script to disable and enable clipmenu every time when `pass` is invoked is clumsy. The password manager already has the feature to delete the latest clipboard entry, and support can be added for clipmenu trivially if deleting the latest entry is available as a feature :)

**Comment by N-R-K** [comment]
> The password manager already has the feature to delete the latest clipboard entry

I don't use `pass` but going through it's manpage I don't see anything like this mentioned. [This is](https://github.com/zx2c4/password-store/blob/b5e965a838bb68c1227caa2cdd874ba496f10149/src/password-store.sh#L193) the closest thing I found to what you're describing.

If by adding "support for clipmenu" you meant patch the source (as opposed to some config) then why can't it be done currently with `clipdel`? E.g just add the line:

```bash
clipdel -d "$(echo -n "$1")"
```

Below the `qdbus` command and it'd work.<sup>*</sup> In fact, this would be *more* robust than "delete current" because in the 45 second window, the user might've copied something else which would push the password to be no longer the current entry.

*: Only issue is regex escaping since `clipdel` expects a regex, not literal string. Adding a flag like `-F` (similar to `grep`) to match a literal instead of regex could be a nice addition.
